src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.cpp

```diff
@@ -0,0 +1,209 @@
+  weighted_lambda_zero_int = step_function(-char_speeds_int[1]);
+  DataVector& weighted_lambda_zero_ext =
+      get(get<::Tags::TempScalar<3>>(buffer));
+  weighted_lambda_zero_ext = -step_function(char_speeds_ext[1]);
+
+  DataVector& weighted_lambda_plus_int =
+      get(get<::Tags::TempScalar<4>>(buffer));
+  weighted_lambda_plus_int = step_function(-char_speeds_int[2]);
+  DataVector& weighted_lambda_plus_ext =
+      get(get<::Tags::TempScalar<5>>(buffer));
+  weighted_lambda_plus_ext = -step_function(char_speeds_ext[2]);
+
+  DataVector& weighted_lambda_minus_int =
+      get(get<::Tags::TempScalar<6>>(buffer));
+  weighted_lambda_minus_int = step_function(-char_speeds_int[3]);
+  DataVector& weighted_lambda_minus_ext =
+      get(get<::Tags::TempScalar<7>>(buffer));
+  weighted_lambda_minus_ext = -step_function(char_speeds_ext[3]);
+
+  for (size_t storage_index = 0;
+       storage_index < char_speed_v_spacetime_metric_ext.size();
+       ++storage_index) {
+    spacetime_metric_boundary_correction->operator[](storage_index) =
+        weighted_lambda_spacetime_metric_ext *
+            char_speed_v_spacetime_metric_ext[storage_index] -
+        weighted_lambda_spacetime_metric_int *
+            char_speed_v_spacetime_metric_int[storage_index];
+  }
+
+  for (size_t storage_index = 0;
+       storage_index < char_speed_v_spacetime_metric_ext.size();
+       ++storage_index) {
+    (*pi_boundary_correction)[storage_index] =
+        0.5 * (weighted_lambda_plus_ext * char_speed_v_plus_ext[storage_index] +
+               weighted_lambda_minus_ext *
+                   char_speed_v_minus_ext[storage_index]) +
+        weighted_lambda_spacetime_metric_ext *
+            char_speed_constraint_gamma2_v_spacetime_metric_ext[storage_index]
+
+        -
+        0.5 * (weighted_lambda_plus_int * char_speed_v_plus_int[storage_index] +
+               weighted_lambda_minus_int *
+                   char_speed_v_minus_int[storage_index]) -
+        weighted_lambda_spacetime_metric_int *
+            char_speed_constraint_gamma2_v_spacetime_metric_int[storage_index];
+  }
+
+  for (size_t storage_index = 0;
+       storage_index < phi_boundary_correction->size(); ++storage_index) {
+    // Overall minus sign on ext because of normal vector is opposite direction.
+    (*phi_boundary_correction)[storage_index] =
+        0.5 * (weighted_lambda_plus_ext *
+                   char_speed_minus_normal_times_v_plus_ext[storage_index] -
+               weighted_lambda_minus_ext *
+                   char_speed_minus_normal_times_v_minus_ext[storage_index]) +
+        weighted_lambda_zero_ext * char_speed_v_zero_ext[storage_index]
+
+        - 0.5 * (weighted_lambda_plus_int *
+                     char_speed_normal_times_v_plus_int[storage_index] -
+                 weighted_lambda_minus_int *
+                     char_speed_normal_times_v_minus_int[storage_index]) -
+        weighted_lambda_zero_int * char_speed_v_zero_int[storage_index];
+  }
+}
+
```

Comment:
Just a note: I checked that these (and the corresponding python functions) agree with what's in the help text